[DONT MERGE] feat(settings/ips): new page to manage IP pools (fix #1350, fix #988, fix #240)

pdonias opened the pull request.

pdonias committed:
feat(settings/ips): new page to manage IP pools

pdonias committed:
feat(settings/ips): match IP pools with networks and filter SelectIp according to it

pdonias committed:
feat(settings/ips): support IP V4 ranges

pdonias committed:
feat(vm/tab-network): toggle default locking mode

pdonias committed:
feat(settings/ips): add networks to existing IP pool

julien-f commented on src/common/select-objects.js:
```diff
@@ -802,3 +803,69 @@ export class SelectSshKey extends Component {
+
+// ===================================================================
+
+export class SelectIp extends Component {
```
Try to use `makeSubscriptionSelect()`.

julien-f commented on src/common/xo/index.js:
```diff
@@ -852,6 +854,10 @@ export const deleteVif = vif => (
   _call('vif.delete', { id: resolveId(vif) })
 )
 
+export const setVif = ({ vif, allowedIpv4Addresses, allowedIpv6Addresses }) => (
```
Nope, look how the params of other setters are laid out.

julien-f commented on src/common/xo/index.js:
```diff
@@ -1550,3 +1556,25 @@ export const updateSchedule = ({ id, job: jobId, cron, enabled, name, timezone }
+export const createIpPool = ({ name, ips, networks }) => {
+  const addresses = {}
+  forEach(ips, ip => { addresses[ip] = {} })
+  return _call('ipPool.create', { name, addresses, networks })::tap(
```
Maybe use `resolveIds()` for networks.

julien-f commented on src/common/xo/index.js:
```diff
@@ -1550,3 +1556,25 @@ export const updateSchedule = ({ id, job: jobId, cron, enabled, name, timezone }
+}
+
+export const deleteIpPool = id => (
+  _call('ipPool.delete', { id })::tap(
```
`resolveId()`.

julien-f commented on src/common/xo/index.js:
```diff
@@ -1550,3 +1556,25 @@ export const updateSchedule = ({ id, job: jobId, cron, enabled, name, timezone }
+  )
+)
+
+export const setIpPool = ({ id, name, addresses, networks }) => (
```
Params are not set like other methods.

julien-f replied:
Use `resolveIds()` for networks.

julien-f commented on src/icons.scss:
```diff
@@ -680,6 +684,10 @@
+    &-ips {
+      @extend .fa;
+      @extend .fa-map-marker;
+    }
```
Let's stop duplicate code needlessly, use `xo-icon-ip` for that.

julien-f commented on src/xo-app/new-vm/index.js:
```diff
@@ -500,6 +502,23 @@ export default class NewVm extends BaseComponent {
     }
   }
   _onChangeSshKeys = keys => this._setState({ sshKeys: map(keys, key => key.id) })
+  _getOnChangeIps = index => ips => {
```
I really dislike this factories!

Usually it means you should have created a dedicated component.

julien-f commented on src/xo-app/new-vm/index.js:
```diff
@@ -1038,6 +1057,9 @@ export default class NewVm extends BaseComponent {
 
 // INTERFACES ------------------------------------------------------------------
 
+  _getIpPoolPredicate = vifNetwork =>
```
I hate these factories!

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+import { Text } from 'editable'
+import { confirm } from 'modal'
+import { Container, Row, Col } from 'grid'
+import SingleLineRow from 'single-line-row'
```
Keep'em sorted.

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+@injectIntl
+export default class Ips extends BaseComponent {
+  _create = () => {
+    const { name, ips: { value: pattern }, networks } = this.refs
```
Why using uncontrolled components, I thought controlled was better?

pdonias replied:
For such a small static form, I think it's actually easier to use refs since nothing has to be checked while the user is typing. This way I don't have to manage state variables or even debounce all my inputs.

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+      ips: parseIpPattern(pattern),
+      name: name.value,
+      networks: map(networks.value, network => network.id)
+    }).then(() => {
```
Maybe the fields should be disabled during the operation.

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+    })
+  }
+  _deleteAllIpPools = () =>
+    confirm({
```
Nope, all confirms should go in `xo`!

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+  _toggleNewIps = id => {
+    const { showNewIpForm } = this.state
+    this.setState({
+      showNewIpForm: { ...showNewIpForm, [id]: !(showNewIpForm && showNewIpForm[id]) }
```
This is much to complicated, what does it mean?!

pdonias replied:
For each IP pool, you can add new IPs, and each IP pool has its own small form inside its table row. These forms can be shown or hidden so I need to store booleans in the state to know whether to show them or not.
This line toggles the form for the IP pool `id`.
Is there a better way to do it?

julien-f replied:
Dunno, but you can always add a comment or explicit boolean with an understandable name to hold the condition (`isCurrentlyShown = showNewIpForm && shownNewIpForm[id]`).

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+    forEach(parseIpPattern(this.state.newIp[id]), ip => {
+      addresses[ip] = {}
+    })
+    setIpPool({
```
This returns a promise, what are you doing with it?

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+            <Col mediumSize={5}>
+              <strong>{ip}</strong>
+            </Col>
+            <Col mediumSize={6}>{!isEmpty(ipPool.addresses[ip].vifs)
```
Use intermediary variables to avoid such long identifiers `ipPool.addresses[ip].vifs`)!

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+            </Col>
+            <Col mediumSize={6}>{!isEmpty(ipPool.addresses[ip].vifs)
+              ? map(ipPool.addresses[ip].vifs, vifId => {
+                const state = store.getState()
```
Nope, you should almost never do this!

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,304 @@
+        return <Container>
+          {map(ipPool.networks, networkId => <Row>
+            <Col mediumSize={11}>
+              {getObject(state, networkId).name_label}
```
`renderXoItemFromId()` may be the solution here.

pdonias committed:
fix(select-objects): makeSubscriptionSelect should work with xoContainers

pdonias committed:
Multiple fixes

- remove the delete-all-ip-pools button
- locking mode Toggle in host and pool view
- lock icon in VM view to show whether the network is locked or not
- warning icon in VM view to notify if a VIF is on a locked network without any allowed IP
- better placeholder for IP pool creation

pdonias committed:
Fix: use connectStore instead of getState

pdonias committed:
Disable IP pools for non Premium. Minor fix.

pdonias committed:
Remove onChange factories in VM creation Interface section

julien-f commented on src/common/base-component.js:
```diff
@@ -1,10 +1,32 @@
+  return object
+}
+
+const deepFind = (object, path, depth = 0) => {
```
You can use `lodash/get`.

julien-f commented on src/xo-app/new-vm/index.js:
```diff
@@ -1039,48 +1040,25 @@ export default class NewVm extends BaseComponent {
-              </Button>
-            </Item>
-          </LineItem>
+          <Interface
```
If not too big, re-inline it (because the goal to cache the handlers is already achieved because of `linkState`).

julien-f commented on package.json:
```diff
@@ -71,6 +71,8 @@
     "gulp-uglify": "^2.0.0",
     "gulp-watch": "^4.3.5",
     "human-format": "^0.6.0",
+    "ip-range-generator": "^1.0.1",
```
You need to reimplement the part you're using directly in xo-web as we do not currently support ES6 external deps.

pdonias committed:
Remove autoFocus from DebounceInput

pdonias committed:
Inline Interface component. Better linkState.

pdonias committed:
fix(settings/ips): IpsCell and NetworksCell stateful components

pdonias committed:
fix(common/ip): copy code from ip-range-generator package

julien-f commented on src/common/select-objects.js:
```diff
@@ -268,13 +297,28 @@ const makeSubscriptionSelect = (subscribe, props) => (
-      }
 
-      this._getFilteredXoObjects = createFilter(
+      this._getFilteredXoContainers = createFilter(
```
Just try to handle `predicate` in each instances.

julien-f commented on src/common/select-objects.js:
```diff
@@ -802,3 +846,25 @@ export class SelectSshKey extends Component {
+        label: ip
+      }))
+    )
+    const xoContainers = map(sortBy(ipPools, 'name'), ipPool => ({
```
Don't sort `ipPools` twice.

julien-f commented on src/common/tooltip/index.js:
```diff
@@ -62,7 +62,9 @@ export class TooltipViewer extends Component {
 export default class Tooltip extends Component {
   componentDidMount () {
-    this._addListeners()
+    if (this.props.content) {
```
Remove this changes.

julien-f commented on src/xo-app/host/index.js:
```diff
@@ -123,6 +125,7 @@ const isRunning = host => host && host.power_state === 'Running'
       pifs: getPifs(state, props),
       pool: getPool(state, props),
       srs: getSrs(state, props),
+      vifsByNetwork: getVifsByNetwork(state, props),
```
Migrate this entry directly to the tab.

julien-f commented on src/xo-app/new-vm/index.js:
```diff
@@ -78,7 +80,16 @@ const NB_VMS_MAX = 100
 /* eslint-disable camelcase */
 
-// Sub-components --------------------------------------------------------------
+const getObject = createGetObject((_, id) => id)
```
No need to specify an id selector because it's already the default value :)

julien-f replied:
Ignore this, I was wrong.

julien-f commented on src/xo-app/new-vm/index.js:
```diff
@@ -78,7 +80,16 @@ const NB_VMS_MAX = 100
+
+// Sub-components
+
+const Item = ({ label, children, className }) => (
```
Keep the previous ordering to minimize changes.

julien-f commented on src/xo-app/pool/index.js:
```diff
@@ -53,6 +53,8 @@ import TabStorage from './tab-storage'
     )
   ).sort()
 
+  const getVifsByNetwork = createGetObjectsOfType('VIF').groupBy('$network')
```
Move into the tab.

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,329 @@
+    networks: getNetworks(state, props),
+    vifs: getVifs(state, props)
+  })
+})
```
You could directly do:

``` js
@connectStore(() => ({
  networks: createGetObjectsOfType('network').groupBy('id'),
  vifs: createGetObjectsOfType('VIF').groupBy('id')
}))
```

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,329 @@
+    setIpPool(this.props.ipPool.id, { addresses: toBeRemoved })
+  }
+
+  _toggleNewIps = () => this.setState({ showNewIpForm: !this.state.showNewIpForm })
```
Use `BaseComponent#toggleState()`.

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,329 @@
+  }
+
+  _toggleNewIps = () => this.setState({ showNewIpForm: !this.state.showNewIpForm })
+  _onChangeNewIps = newIps => this.setState({ newIps })
```
Use `BaseComponent#linkState()`.

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,329 @@
+  _onChangeNewNetworks = newNetworks =>
+    this.setState({ newNetworks: map(newNetworks, network => network.id) })
+  _networkPredicate = network =>
+    !includes(this.props.ipPool.networks, network.id)
```
If the predicate has some dependencies they should be explicit by regenerating the predicate `createSelector()`.

julien-f commented on src/xo-app/settings/ips/index.js:
```diff
@@ -0,0 +1,329 @@
+                multi
+                onChange={networks => this._onChangeNewNetworks(networks)}
+                predicate={this._networkPredicate}
+                value={newNetworks || []}
```
Why not injecting `[]` in the state at creation time?

julien-f commented on src/xo-app/vm/tab-network.js:
```diff
@@ -192,13 +297,11 @@ export default class TabNetwork extends Component {
-                                  handlerParam={vif}
-                                />
-                              }
+                              <ActionRowButton
```
Why has this condition disappeared?

pdonias committed:
Multiple fixes.
